on_the_porch/sql_chat/app4.py

```
@@ -0,0 +1,1020 @@
+    for attempt_idx in range(1, max_attempts + 1):
+        try:
+            if attempt_idx == 1:
+                print("\n[SQL]\n" + sql + "\n")
+            else:
+                print(f"\n[SQL Retry {attempt_idx - 1}]\n" + sql + "\n")
+            
+            # Normalize SQL for comparison (remove extra whitespace)
+            sql_normalized = " ".join(sql.split())
+            if sql_normalized in previous_sqls:
+                # Same SQL as before - avoid infinite loop, return early
+                print(f"\n[Warning] SQL same as previous attempt, stopping to avoid infinite loop\n")
```

Ruff (0.14.5)

f-string without any placeholders

Remove extraneous `f` prefix

(F541)